chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,467 @@
+from chainer.functions.array.reshape import reshape
+from chainer.functions.array.stack import stack
+from chainer.functions.array.transpose import transpose
+from chainer.functions.math.average import average
+from chainer.functions.math.basic_math import absolute
+from chainer.functions.math.maximum import maximum
+from chainer.functions.math.sum import sum as c_sum
+import warnings
+
+
+class DiscriminativeMarginBasedClusteringLoss(object):
+    """Discriminative margin based clustering loss function
+
+    This is the implementation of following paper:
+    https://arxiv.org/pdf/1802.05591.pdf
+
+    In segmentation, one of the biggest problem is having noise at the output
+    of trained network.
+    For cross-entropy based approaches, if the pixel value is wrong,
+    the loss value will be same independent from wrong pixel's location.
+    However, for segmentation, even though network gives wrong pixel output,
+    it is desirable to have it
+    as close as possible to the original position.
+    By applying discriminative loss function, groups of segmentation instances
+    can be moved together.
+
+    This loss function calculates three different parameters:
+        - Variance Loss:
+            Loss to penalize distances between pixels which are belonging
+            to same instance. (Pull force)
+        - Distance loss:
+            Loss to penalize distances between the centers of instances.
+            (Push force)
+        - Regularization loss:
+            Small regularization loss to penalize weights against overfit
+
+    Args:
+        delta_v (float): Minimum distance to start penalizing variance
+        delta_d (float): Maximum distance to stop penalizing distance
+        max_n_clusters (int): Maximum possible number of clusters.
+        norm (int): Norm to calculate pixels and cluster center distances
+        alpha (float): Weight for variance loss      (alpha * variance_loss)
+        beta (float): Weight for distance loss       (beta * distance_loss)
+        gamma (float): Weight for regularization loss(gamma * regularizer_loss)
+
+    Returns:
+        ~chainer.Variable: Output variable. A
+        :math:`()`-shaped float array.
+    """
+
+    def __init__(self, delta_v, delta_d, max_n_clusters, norm=1, alpha=1.0,
+                 beta=1.0, gamma=0.001):
+        self.delta_v = delta_v
+        self.delta_d = delta_d
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+        self.max_n_clusters = max_n_clusters
+
+        # L1 or L2 norm is allowed only
+        if norm == 1:
+            self.norm = self._l1_norm
+        elif norm == 2:
+            self.norm = self._l2_norm
+        else:
+            raise ValueError("Norm can only be 1 or 2")
+
+    def _l1_norm(self, x, axis):
+        """Function to calculate L1 Norm by given axes
+
+        Args:
+            x (:class:`~chainer.Variable\` or \
+            :class:\`numpy.ndarray\` or \
```

Comment:
Are back slashes of backquotes intentional?

Reply:
Yes, they are intentional. Do you suggest to remove?

Reply:
I'm just curious why you did so because you didn't put back slashes in other places.

Reply:
If it is a multi-line explanation of the type, I appended backslashes for easier readability. If it is a single line explanation, I did not add. As the only multi-line type explanation is :class:`~chainer.Variable\` or \
:class:\`numpy.ndarray\` or \
:class:`cupy.ndarray` 
I used backslashes only for those.

Reply:
What I meant was you added back slashes to backquotes of `numpy.ndarray` (i.e. `` \`numpy.ndarray\` ``) while you do not those of `cupy.ndarray` (i.e. `` `cupy.ndarray` ``) .

Reply:
Oh, now got it. That was a mistake. I cleaned them. Thanks for catching.